File: LeetCode/Medium/0621-task-scheduler/0621-task-scheduler.java
class Solution {
    public int leastInterval(char[] tasks, int n) {
        int[] map = new int[26];
        for (char task : tasks)
            map[task - 'A']++;

        int maxCount = 0, numOfMaxCount = 0;
        for (int num : map) {
            if (num > maxCount) {
                maxCount = num;
                numOfMaxCount = 1;
            } else if (num == maxCount) {
                numOfMaxCount++;
            }
        }

        int res = (n + 1) * (maxCount - 1) + numOfMaxCount;

        return Math.max(res, tasks.length);
    }
}

File: LeetCode/Medium/0621-task-scheduler/0621-task-scheduler.kt
class Solution {
    fun leastInterval(tasks: CharArray, n: Int): Int {
        val freq = IntArray(26)
        for (ch in tasks) {
            freq[ch.code - 'A'.code]++
        }

        val pq: PriorityQueue<Int> = PriorityQueue<Int>(Collections.reverseOrder())
        for (i in 0..25) {
            if (freq[i] > 0) {
                pq.offer(freq[i])
            }
        }

        var time = 0
        while (!pq.isEmpty()) {
            var cycle = n + 1
            val store: MutableList<Int> = ArrayList()
            var taskCount = 0
            while (cycle-- > 0 && !pq.isEmpty()) {
                val currentFreq = pq.poll()
                if (currentFreq > 1) {
                    store.add(currentFreq - 1)
                }
                taskCount++
            }
            store.forEach(Consumer { e: Int -> pq.offer(e) })
            time += (if (pq.isEmpty()) taskCount else n + 1)
        }
        return time
    }
}